### PR TITLE
Rules generation

### PR DESCRIPTION
**Pull request by riccardocadei**
Consider also internal nodes for rules generation.
Add max_rules filtering in rules generation,
Simplify rules generation hyperparameters.

### CODE REVIEW

File: R/predict.R [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#'
+#' @param CRE A cre object from running the CRE function.
+#' @param X A covariate matrix (or data.frame)
+#'
+#' @return
+#' Array with the estimated Individual Treatment Effects
+#'
+#' @export
+#'
+predict.cre <- function(CRE, X) {
```
Comment (Naeemkh) [inline R/predict.R:15]: In generic functions use `object` for consistency. This is section fails the tests.

**Review by Naeemkh — CHANGES_REQUESTED** [review]

File: R/predict.R (hunk 1 above)
Comment (riccardocadei) [reply R/predict.R:15]: Oh good to know, thank you!

File: NEWS.md [hunk 2]
```diff
@@ -1,13 +1,45 @@
...
+* `rules`(implicit form) in cre() function return.
+* predict() function for ITE estimation via CRE.
+
+## Changed 
+* Type `stability_selection` binary -> string ('no','vanilla','error_control').
+* Unify `ntrees_gbm` hyper-parameter and `ntrees_gbm` hyper-parameter in 
+`ntrees` hyper-parameter.
+* In rules generation retrieve decision rules also from internal nodes, and not
+just from terminal nodes.
+* `ite_method_dis`, `ite_method_inf` method-parameter -> `ite_method`.
```
Comment (Naeemkh) [inline NEWS.md:18]: Please also update the readme file. It is still based on `ite_method_dis` and `ite_method_inf`.

File: R/cre.R [hunk 3]
```diff
@@ -188,11 +182,13 @@ cre <- function(y, z, X,
 
   # Estimate ITE
   if (is.null(ite)) {
-    ite_inf <- estimate_ite(y = y_inf, z = z_inf, X = X_inf,
-                    ite_method = getElement(method_params, "ite_method_inf"),
-                    ps_method = getElement(method_params, "ps_method_inf"),
-                    oreg_method = getElement(method_params, "oreg_method_inf"),
-                    offset = getElement(method_params, "offset"))
+    ite_inf <- estimate_ite(y = y_inf,
+                          z = z_inf,
```
Comment (Naeemkh) [inline R/cre.R:186]: Please fix the indentation.

File: R/predict.R [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#' @title
+#' Predict Individual Treatment Effect via Causal Rule Ensemble
```
Comment (Naeemkh) [inline R/predict.R:2]: Only first word should be capitalized.

File: R/predict.R [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#'
+#' @description
+#' Predict Individual Treatment Effect via Causal Rule Ensemble algorithm.
```
Comment (Naeemkh) [inline R/predict.R:5]: Also this one. In description, it should be `Predicts`

File: R/predict.R [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+#'
+#' @param object A cre object from running the CRE function.
+#' @param X A covariate matrix (or data.frame)
+#' @param ... Additional arguments passed to customize the prediction.
+#'
+#' @return
+#' Array with the estimated Individual Treatment Effects
```
Comment (Naeemkh) [inline R/predict.R:12]: An array

File: R/print.R [hunk 7]
```diff
@@ -52,61 +51,48 @@ summary.cre <- function(object, verbose = 2, ...) {
...
-    cat("\n    - Number of Trees       :", getElement(params, "ntrees_rf"),
-        "RF +", getElement(params, "ntrees_gbm"), "GBM")
-    cat("\n    - Node Size             :", getElement(params, "node_size"))
-    cat("\n    - Max Nodes             :", getElement(params, "max_nodes"))
-    cat("\n    - Max Depth             :", getElement(params, "max_depth"))
-    cat("\n  - Filtering")
-    cat("\n    - Threshold Decay (Irrelevant):", getElement(params,
+    cat("\n  - Number of Trees       :", getElement(params, "ntrees"))
+    cat("\n  - Node Size             :", getElement(params, "node_size"))
+    #cat("\n  - Max Rules             :", getElement(params, "max_rules"))
```
Comment (Naeemkh) [inline R/print.R:76]: Please remove commented codes.

File: R/print.R [hunk 8]
```diff
@@ -52,61 +51,48 @@ summary.cre <- function(object, verbose = 2, ...) {
...
-    cat("\n- CATE Inference")
-    if (getElement(params, "ite_method_inf") == "tpoisson") {
-      cat("\n    - Estimator       : tlearner")
-      cat("\n    - Outcome         : poisson")
-      cat("\n    - Offset          :", getElement(params, "offset"))
-    } else {
-      cat("\n    - Estimator       :", getElement(params, "ite_method_inf"))
-      cat("\n    - Outcome         :", getElement(params, "oreg_method_inf"))
-      cat("\n    - Propensity Score:", getElement(params, "ps_method_inf"))
+    if (getElement(params, "stability_selection")=="error_control") {
```
Comment (Naeemkh) [inline R/print.R:87]: You can get the element once. Also, please follow tidyverse styling guide.

```r
stability_selection <- getElement(params, "stability_selection")

if (stability_selection == "error_control") {
  cat("\n- Stability Selection with Error Control")
  cat("\n  - Cutoff:", getElement(params, "cutoff"))
  cat("\n  - PFER  :", getElement(params, "pfer"))
} else if (stability_selection == "vanilla") {
  cat("\n- Vanilla Stability Selection")
  cat("\n  - Cutoff:", getElement(params, "cutoff"))
} else if (stability_selection == "no") {
  cat("\n- No Stability Selection")
}

```

File: R/select_rules.R [hunk 9]
```diff
@@ -7,46 +7,63 @@
 #' Stability Selection).
 #'
 #' @param rules_matrix The rules matrix.
-#' @param rules_list A list of rules.
+#' @param rules A list of rules.
```
Comment (Naeemkh) [inline R/select_rules.R:10]: For consistency in documentation, always choose either `A` or `The`. My recommendation is to use `A` throughout.

File: R/select_rules.R [hunk 10]
```diff
@@ -7,46 +7,63 @@
...
+        indices <- sample(1:nrow(R),
+                          size = round(nrow(R) * subsample),
+                          replace = FALSE)
+        lasso <- glmnet::cv.glmnet(x = as.matrix(R[indices, ]),
+                                   y = ite[indices] - ite_mean,
+                                   alpha = 1,
+                                   nfolds = 5,
+                                   gamma = c(0.01, 0.05, 0.1, 0.5, 1, 5, 10),
+                                   intercept = FALSE)
+        non_zero_indices <- which(as.matrix(coef(lasso)) != 0)-1
```
Comment (Naeemkh) [inline R/select_rules.R:56]: Styling: space before and after -

File: R/select_rules.R [hunk 11]
```diff
@@ -7,46 +7,63 @@
...
+        stability_scores[non_zero_indices] <- stability_scores[non_zero_indices]+1
```
Comment (Naeemkh) [inline R/select_rules.R:57]: Styling: space before and after +

File: tests/testthat/test-cre.R [hunk 12]
```diff
@@ -10,125 +10,94 @@ test_that("cre Runs Correctly", {
...
+                        learner_y = "SL.xgboost")
 
  hyper_params <- list(intervention_vars = NULL,
                       offset = NULL,
-                      ntrees_rf = 20,
-                      ntrees_gbm = 20,
+                      ntrees = 50,
                       node_size = 10,
-                      max_nodes = 5,
+                      max_rules = 50,
```
Comment (Naeemkh) [inline tests/testthat/test-cre.R:21]: Please make sure that the code is comprehensively tested. Any alterations to test values necessitate a meticulous review.

**Review by Naeemkh — CHANGES_REQUESTED** [review]

**Pull request merged**